src/chainstate/stacks/db/blocks.rs

```diff
@@ -2702,6 +2737,107 @@ impl StacksChainState {
+           fee / tx_size < MINIMUM_TX_FEE_RATE_PER_BYTE {
+            return Err(MemPoolRejection::FeeTooLow(fee, cmp::max(MINIMUM_TX_FEE, tx_size * MINIMUM_TX_FEE_RATE_PER_BYTE)))
+        }
+
+        // 4: the account nonces must be correct
+        let (origin, payer) = self.with_read_only_clarity_tx(current_burn, current_block, |conn| {
+            StacksChainState::check_transaction_nonces(conn, &tx)
+                .map_err(|e| MemPoolRejection::BadNonces(e))
+        })?;
+
+        // 5: the paying account must have enough funds
+        if fee as u128 > payer.stx_balance {
+            return Err(MemPoolRejection::NotEnoughFunds(fee as u128, payer.stx_balance))
+        }
+
+        // 6: payload-specific checks
+        match &tx.payload {
+            TransactionPayload::TokenTransfer(_addr, amount, _memo) => {
+                // got the funds?
+                let total_spent = (*amount as u128) +
+                    if origin == payer {
+                        fee as u128
+                    } else {
+                        0
+                    };
+                if total_spent > origin.stx_balance {
+                    return Err(MemPoolRejection::NotEnoughFunds(total_spent, origin.stx_balance))
+                }
+            },
+            TransactionPayload::ContractCall(TransactionContractCall {
+                address, contract_name, function_name, function_args }) => {
+                let contract_identifier = QualifiedContractIdentifier::new(address.clone().into(), contract_name.clone());
+
+                let function_type = self.with_read_only_clarity_tx(current_burn, current_block, |conn| {
+                    conn.with_analysis_db_readonly(|db| {
+                        db.get_public_function_type(&contract_identifier, &function_name)
+                    }) })
+                    .map_err(|_e| MemPoolRejection::NoSuchContract)?
+                    .ok_or_else(|| MemPoolRejection::NoSuchPublicFunction)?;
+
+                let arg_types: Vec<_> = function_args.iter().map(|x| TypeSignature::type_of(x)).collect();
+                function_type.check_args(&arg_types)
+                    .map_err(|e| MemPoolRejection::BadFunctionArgument(e))?;
+            },
+            TransactionPayload::SmartContract(TransactionSmartContract { name, code_body: _ }) => {
+                let contract_identifier = QualifiedContractIdentifier::new(tx.origin_address().into(), name.clone());
+                let exists = self.with_read_only_clarity_tx(current_burn, current_block, |conn| {
+                    conn.with_analysis_db_readonly(|db| {
+                        db.has_contract(&contract_identifier)
+                    }) });
+                if exists {
+                    return Err(MemPoolRejection::ContractAlreadyExists(contract_identifier))
+                }
+            },
+            TransactionPayload::PoisonMicroblock(microblock_header_1, microblock_header_2) => {
```

Comment:
To be clear, the code for processing a poison microblock transaction isn't written yet (it'll cause a panic).

Reply:
Yep, though we can still implement the admission check.